TSPL.docc/ReferenceManual/Declarations.md

```diff
@@ -980,6 +1040,152 @@ see <doc:Functions#In-Out-Parameters>.
   ```
 -->
 
+#### Borrowing and Consuming Parameters
+
+By default, Swift uses a set of rules
+to automatically manage manage object lifetime across function calls,
+copying values when required.
+The default rules are designed to minimize overhead in most cases ---
+if you want more specific control,
+you can apply the `borrowing` or `consuming` parameter modifier.
+In this case,
+you must use `copy` to explicitly mark copy operations.
+
+Regardless of whether you use the default rules,
+Swift guarantees that object lifetime and
+ownership are correctly managed in all cases.
+These parameter modifiers impact only the relative efficiency
+of particular usage patterns, not correctness.
+
+<!-- XXX TR:
+Should we describe the default rules somewhere?
+Here or in the ARC chapter?
+-->
+
+The `borrowing` modifier indicates that the function
+does not keep the parameter's value.
+In this case, the caller maintains ownership of the object
+and the responsibility for the object's lifetime.
+Using `borrowing` minimizes overhead when the function
+is making only transient use of the object.
+
+```swift
+// `isLessThan` does not keep either argument
+func isLessThan(lhs: borrowing A, rhs: borrowing A) -> Bool {
+    ...
+}
+```
+
+If the function needs to keep the parameter's value
+for example, by storing it in a global variable ---
+you use `copy` to explicitly copy that value.
+
+```swift
+// As above, but this `isLessThan` also wants to record the smallest value
+func isLessThan(lhs: borrowing A, rhs: borrowing A) -> Bool {
+    if lhs < storedValue {
+        storedValue = copy lhs
+    } else if rhs < storedValue {
+        storedValue = copy rhs
+    }
+    return lhs < rhs
+}
+```
+
+Conversely,
+the `consuming` parameter modifier indicates
+that the function takes ownership of the value,
+accepting responsibility for either storing or destroying it
+before the function returns.
+
+```swift
+// `store` keeps its argument, so we mark it `consuming`
+func store(a: consuming A) {
+    someGlobalVariable = a
+}
+```
+
+Using `consuming` minimizes overhead when the caller no longer
+needs to use the object after the function call.
+
+```swift
+// Usually, this is the last thing we do with a value
+store(a: value)
+```
+
+If the caller does need to use the object after the function call,
+it might need to make a separate
+copy before calling the function.
+<!-- XXX TR:
+When do you need to explicitly copy?
+When does the compiler implicitly insert a copy, as shown below?
+-->
```

Comment:
@tbkka When does copying need to be explicit, and when is it implicit?

Reply:
I believe that technically, the compiler today always inserts a copy (unless the type is not copyable) and then relies on the optimizer to remove the copy if it turns out to be redundant.  However, I'm not sure we should document this behavior as a requirement of the Swift language.  We should certainly not document it here, as it's really a feature of Swift's general function-calling convention, which is independent of these particular modifiers.

Reply:
Thanks, Tim — removed the XXX marker in e738f1a0, so this doesn't block merging.

Can we rewrite the "might need" bit in the sentence?  As is, it's unclear when you do and don't need to make a copy.

Reply:
Apologies, I think I just realized where we're mis-communicating.  I was thinking of this from the compiler's point of view.  It should probably read:

> If the caller does need to use the object after the function call, the compiler will automatically make a copy of the object if it can.  If the object is non-copyable, the compiler will not be able to do this for you.

Reply:
👍 Revised in e98ee058